bokehjs/test/models/mappers/categorical_color_mapper.ts

```diff
@@ -0,0 +1,342 @@
+          expect(vals0).to.be.deep.equal(["gray"])
+
+          const vals1 = cm0.v_compute([["a", "1", "baz"]])
+          expect(vals1).to.be.deep.equal(["gray"])
+
+          const cm1 = new CategoricalColorMapper({
+            palette,
+            factors: factors.map((x) => x.slice(1, 2)[0]),
+            start: i,
+            end: j,
+          })
+
+          const vals2 = cm1.v_compute([["a", "1", "foo"]])
+          expect(vals2).to.be.deep.equal(["gray"])
+
+          const vals3 = cm1.v_compute([["a", "1", "baz"]])
+          expect(vals3).to.be.deep.equal(["gray"])
+
+          const cm2 = new CategoricalColorMapper({
+            palette,
+            factors: factors.map((x) => x.slice(2, 3)[0]),
+            start: i,
+            end: j,
+          })
+
+          const vals4 = cm2.v_compute([["a", "1", "foo"]])
+          expect(vals4).to.be.deep.equal(["gray"])
+
+          const vals5 = cm2.v_compute([["a", "1", "baz"]])
+          expect(vals5).to.be.deep.equal(["gray"])
+        })
+      }
+    })
+
+    describe("and 2-level palette factors", () => {
+
+      it("should map factors to palette with start=0, end=2", () => {
+        const palette = ["red", "green", "blue", "orange"]
+        const factors: F3[] = [["a", "1", "foo"], ["a", "2", "foo"], ["b", "2", "baz"], ["c", "1", "bar"]]
+        const cm = new CategoricalColorMapper({
+          palette,
+          factors: factors.map((x) => x.slice(0, 2) as F2),
+          end: 2,
+        })
+        const vals = cm.v_compute(factors)
+        expect(vals).to.be.deep.equal(palette)
+      })
+
+      it("should map factors to palette with start=1, end=3", () => {
+        const palette = ["red", "green", "blue", "orange"]
+        const factors: F3[] = [["a", "1", "foo"], ["a", "2", "foo"], ["b", "2", "baz"], ["c", "1", "bar"]]
+        const cm = new CategoricalColorMapper({
+          palette,
+          factors: factors.map((x) => x.slice(1, 3) as F2),
+          start: 1,
+          end: 3,
+        })
+
+        const vals = cm.v_compute(factors)
+        expect(vals).to.be.deep.equal(palette)
+      })
+
+      for (const [i, j] of [[0, 1], [0, 3], [1, 2], [2, 3]]) {
+        it(`should map everything to nan_color with start=${i}, end=${j}`, () => {
+          const palette = ["red", "green", "blue", "orange"]
+          const factors: F3[] = [["a", "1", "foo"], ["a", "2", "foo"], ["b", "2", "baz"], ["c", "1", "bar"]]
+
+          const cm0 = new CategoricalColorMapper({
+            palette,
+            factors: factors.map((x) => x.slice(0, 2) as F2),
+            start: i,
+            end: j,
+          })
+
+          const vals0 = cm0.v_compute(["a"])
+          expect(vals0).to.be.deep.equal(["gray"])
+
+          /* XXX: gives ["red"] because original code used [i, j] = [3, 4] due to lack of lexical scoping
+          const vals1 = cm0.v_compute([["a", "1"]])
+          expect(vals1).to.be.deep.equal(["gray"])
+          */
```

Comment:
@bryevdv, the original code always used `[i, j] = [3, 4]` due to CoffeeScript's messed up scoping rules (`for` loop without `do` statement). This now fails with `["red"]` result, but I thing this is actually correct behavior.

Reply:
I concur it should be `"red"`